README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# HohgantR <img src="man/figures/logo.png" align="right" height="138" />

<!-- badges: start -->

[![forthebadge](https://forthebadge.com/images/badges/works-on-my-machine.svg)](https://forthebadge.com)
[![](https://img.shields.io/badge/devel%20version-0.0.1-blue.svg)](https://github.com/BBieri/HohgantR)
[![](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

The goal of HohgantR is to serve as a toolbox filled with all the
miscellaneous functions that might be useful one day, or not…

## Installation

You can install the latest version of HohgantR from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("BBieri/HohgantR")
```

A release on CRAN is not planned as the package is intended to serve as
a “cabinet de curiosités” of R functions.

## Slope charts à la Tufte

A `{ggplot2}` graphing function inspired by and updated from [this great
package](https://github.com/leeper/slopegraph).

``` r
library(HohgantR)
HohgantR::indicators_long |>
dplyr::filter(indicator == "Indicator") |> # Filter by indicator if necessary
ggslope(year, value, unit)  # Plot things
```

<img src="man/figures/README-ggslopeexample-1.png" width="100%" />

## Bernie’s theme

A `{ggplot2}` theme for Bernie’s blog.

``` r
library(HohgantR)
library(tidyverse)
#> Warning in Sys.timezone(): unable to identify current timezone 'C':
#> please set environment variable 'TZ'
# Prepare data
cty_mpg <- aggregate(mpg$cty, by = list(mpg$manufacturer), FUN = mean)
colnames(cty_mpg) <- c("make", "mileage") # change column names
cty_mpg <- cty_mpg[order(cty_mpg$mileage), ] # sort
cty_mpg$make <- factor(cty_mpg$make, levels = cty_mpg$make)
# Plot
ggplot(cty_mpg, aes(x = make, y = mileage)) +
  geom_segment(aes(
    x = make,
    xend = make,
    y = 0,
    yend = mileage
  ), color = "white") +
  geom_point(size = 3, color = "orange") +
  labs(
    title = "Lollipop Chart",
    subtitle = "Make Vs Avg. Mileage",
    caption = "source: mpg"
  ) +
  HohgantR::themebernie() +
  theme(axis.text.x = element_text(angle = 65, vjust = 0.6))
```

<img src="man/figures/README-berniestheme-1.png" width="100%" />

## Mineral theme

A `{ggplot2}` theme inspired by minerals.

``` r
library(HohgantR)
library(tidyverse)
# Prepare data
cty_mpg <- aggregate(mpg$cty, by = list(mpg$manufacturer), FUN = mean)
colnames(cty_mpg) <- c("make", "mileage") # change column names
cty_mpg <- cty_mpg[order(cty_mpg$mileage), ] # sort
cty_mpg$make <- factor(cty_mpg$make, levels = cty_mpg$make)
# Plot
ggplot(cty_mpg, aes(x = make, y = mileage)) +
  geom_segment(aes(
    x = make,
    xend = make,
    y = 0,
    yend = mileage
  ), color = "white") +
  geom_point(size = 3, color = "#C3CED6") +
  labs(
    title = "Lollipop Chart",
    subtitle = "Make Vs Avg. Mileage",
    caption = "source: mpg"
  ) +
  HohgantR::thememineral() +
  theme(axis.text.x = element_text(angle = 65, vjust = 0.6))
```

<img src="man/figures/README-mineraltheme-1.png" width="100%" />


NEWS.md
# HohgantR 0.0.1

## Package

* Added `{pkgdown}` site for documentation purposes.

## `{ggplot2}` themes

* Added `themebernie()`, a dark theme for Bernie's blog.
* Added `thememineral()`,, a dark theme with mineral colors

## `{ggplot2}` graphs

* Added `ggslope()` an updated adaptation from
[this great package](https://github.com/leeper/slopegraph)

# HohgantR 0.0.0.9000

* Initialized the package.
* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(HohgantR)

testthat::test_check("HohgantR")


tests/testthat/test-ggslopegraph.R
testplot <- HohgantR::indicators_long |>
  dplyr::filter(indicator == "Indicator") |>
  ggslope(year, value, DisaggregationLevel)

test_that("ggslope() works", {
  expect_equal(class(testplot[[1]]), c("tbl_df", "tbl", "data.frame"))
  expect_equal(length(testplot), 9)
})


R/ggslopegraph.R
#' Slope graph for ggplot2
#'
#' Creates a slope graph ideal to use when computing changes in one or more
#' indicators between waves of a survey or other time series data with
#' a low number of periods.
#'
#' @param data Data frame containing the data in a long form. Required input.
#' @param x The variable on the x axis, usually time, survey wave, etc. Required
#' input.
#' @param y The variable on the y axis, usually an indicator such as poverty and
#' inequality measures. Required input.
#' @param group Grouping variable showing various levels of the indicator.
#' @param DataLabel Data label
#' @param title Plot title
#' @param subtitle Plot subtitle
#' @param caption Plot caption
#' @param LineThickness Line width
#' @param LineColor Line color
#' @param DataTextSize Text size of the data point value.
#' @param DataTextColor Text color of the data point value.
#' @param DataLabelPadding Text padding of the label of each level.
#' @param DataLabelLineSize Text label line size.
#' @param DataLabelFillColor Text label fill color.
#' @param WiderLabels Set wider labels? False by default.
#' @param ReverseYAxis Revert Y axis? False by default.
#' @param ReverseXAxis Revert X axis? False by default.
#' @param RemoveMissing Remove missing values? True by default.
#'
#' @source Adapted and updated from
#' [this great package](https://github.com/leeper/slopegraph).
#' @return A pretty slope graph.
#' @export
#'
#' @examples
#' HohgantR::indicators_long |>
#' dplyr::filter(indicator == "Indicator") |>
#' ggslope(year, value, unit)
#'

ggslope <- function(data,
                    x,
                    y,
                    group,
                    DataLabel = NULL,
                    title = "No title given",
                    subtitle = "No subtitle given",
                    caption = "No caption given",
                    LineThickness = 1,
                    LineColor = "ByGroup",
                    DataTextSize = 2.5,
                    DataTextColor = "black",
                    DataLabelPadding = 0.05,
                    DataLabelLineSize = 0,
                    DataLabelFillColor = "#f5f5f2",
                    WiderLabels = FALSE,
                    ReverseYAxis = FALSE,
                    ReverseXAxis = FALSE,
                    RemoveMissing = TRUE) {

  # ---------------- input checking ----------------------------

  # Init .data to avoid note
  .data <- NULL

  # error checking and setup
    if (length(match.call()) <= 4) {
      stop("Not enough arguments passed requires a data, plus at least three variables")
    }
    argList <- as.list(match.call()[-1])
    if (!methods::hasArg(data)) {
      stop("You didn't specify a data to use", call. = FALSE)
    }

    Nx <- deparse(substitute(x)) # name of x variable
    Ny <- deparse(substitute(y)) # name of y variable
    Ngroup <- deparse(substitute(group)) # name of group variable

    if (is.null(argList$DataLabel)) {
      NDataLabel <- deparse(substitute(y))
      DataLabel <- argList$y
    } else {
      NDataLabel <- deparse(substitute(DataLabel))
    }

  Ndata <- argList$data # name of data
  if (!methods::is(data, "data.frame")) {
    stop(paste0("'", Ndata, "' does not appear to be a |> frame"))
  }
  if (!Nx %in% names(data)) {
    stop(paste0("'", Nx, "' is not the name of a variable in the data"),
         call. = FALSE)
  }
  if (anyNA(data[[Nx]])) {
    stop(paste0("'", Nx, "' can not have missing data please remove those rows!"),
         call. = FALSE)
  }
  if (!Ny %in% names(data)) {
    stop(paste0("'", Ny, "' is not the name of a variable in the data"),
         call. = FALSE)
  }
  if (!Ngroup %in% names(data)) {
    stop(paste0("'", Ngroup, "' is not the name of a variable in the data"),
         call. = FALSE)
  }
  if (!NDataLabel %in% names(data)) {
    stop(paste0("'", NDataLabel, "' is not the name of a variable in the data"),
         call. = FALSE)
  }
  if (anyNA(data[[Ngroup]])) {
    stop(paste0(
      "'",
      Ngroup,
      "' can not have missing data please remove those rows!"
    ),
    call. = FALSE)
  }
  if (!class(data[[Ny]]) %in% c("integer", "numeric")) {
    stop(paste0("Sorry I need the measured variable '", Ny, "' to be a number"),
         call. = FALSE)
  }
  if (!"ordered" %in% class(data[[Nx]])) {
    # keep checking
    if (!"character" %in% class(data[[Nx]])) {
      # keep checking
      if ("factor" %in% class(data[[Nx]])) {
        # impose order
        message(paste0("\nConverting '", Nx, "' to an ordered factor\n"))
        data[[Nx]] <- factor(data[[Nx]], ordered = TRUE)
      } else {
        stop(
          paste0(
            "Sorry I need the variable '",
            Nx,
            "' to be of class character, factor or ordered"
          ),
          call. = FALSE
        )
      }
    }
  }

  x <- dplyr::enquo(x)
  y <- dplyr::enquo(y)
  group <- dplyr::enquo(group)
  DataLabel <- dplyr::enquo(DataLabel)

  # Handle edge cases

  if (ReverseXAxis) {
    data[[Nx]] <- forcats::fct_rev(data[[Nx]])
  }

  NumbOfLevels <- nlevels(factor(data[[Nx]]))
  if (WiderLabels) {
    CustomTheme <-
      c(CustomTheme, ggplot2::expand_limits(x = c(0, NumbOfLevels + 5)))
  }

  if (ReverseYAxis) {
    CustomTheme <- c(CustomTheme, ggplot2::scale_y_reverse())
  }

  if (length(LineColor) > 1) {
    if (length(LineColor) < length(unique(data[[Ngroup]]))) {
      message(
        paste0(
          "\nYou gave me ",
          length(LineColor),
          " colors I'm recycling colors because you have ",
          length(unique(data[[Ngroup]])),
          " ",
          Ngroup,
          "s\n"
        )
      )
      LineColor <-
        rep(LineColor, length.out = length(unique(data[[Ngroup]])))
    }
    LineGeom <-
      list(ggplot2::geom_line(ggplot2::aes(color = .data[[group]]), size = LineThickness),
           ggplot2::scale_color_manual(values = data[[LineColor]]))
  } else {
    if (LineColor == "ByGroup") {
      LineGeom <-
        list(ggplot2::geom_line(ggplot2::aes(color = .data[[group]], alpha = 1), size = LineThickness))
    } else {
      LineGeom <-
        list(ggplot2::geom_line(ggplot2::aes(), size = LineThickness, color = LineColor))
    }
  }

  # logic to sort out missing values if any
  if (anyNA(data[[Ny]])) {
    # are there any missing
    if (RemoveMissing) {
      # which way should we handle them
      data <- data |>
        dplyr::group_by(!!group) |>
        dplyr::filter(!anyNA(!!y)) |>
        droplevels()
    } else {
      data <- data |>
        dplyr::filter(!is.na(!!y))
    }
  }

  # The actual graph
  data |>
    ggplot2::ggplot(ggplot2::aes(group = .data[[group]],
                                 x = .data[[x]],
                                 y = .data[[y]])) +
    LineGeom +
    # left side y axis labels
    ggrepel::geom_text_repel(
      data = dplyr::slice_min(data, eval(x), n = 1),
      ggplot2::aes(label = .data[[group]]),
      hjust = "left",
      box.padding = 0.10,
      point.padding = 0.10,
      segment.color = "gray",
      segment.alpha = 0.6,
      fontface = "bold",
      size = 3,
      nudge_x = -0.5,
      direction = "y",
      force = 1.5,
      max.iter = 3000
    ) +
    # right side y axis labels
    ggrepel::geom_text_repel(
      data = dplyr::slice_max(data, eval(x), n = 1),
      ggplot2::aes(label = .data[[group]]),
      hjust = "right",
      box.padding = 0.10,
      point.padding = 0.10,
      segment.color = "gray",
      segment.alpha = 0.6,
      fontface = "bold",
      size = 3,
      nudge_x = 0.5,
      direction = "y",
      force = 1.5,
      max.iter = 3000
    ) +
    # data point labels
    ggplot2::geom_label(ggplot2::aes(label = .data[[NDataLabel]]),
               size = DataTextSize,
               label.padding = ggplot2::unit(DataLabelPadding, "lines"),
               label.size = DataLabelLineSize,
               color = DataTextColor,
               fill = DataLabelFillColor
    ) +
    ggplot2::scale_x_discrete(position = "top") +
    CustomTheme() +
    ggplot2::labs(
      title = title,
      subtitle = subtitle,
      caption = caption
    )
}

# ggslope theme

CustomTheme <- function(){
  ggplot2::theme_minimal() +
    ggplot2::theme(
      legend.position = "none",
      axis.text.x = ggplot2::element_text(size = 12, vjust = 1.5),
      axis.text.y = ggplot2::element_blank(),
      axis.title.x = ggplot2::element_blank(),
      axis.title.y = ggplot2::element_blank(),
      panel.grid.major = ggplot2::element_blank(),
      panel.grid.minor = ggplot2::element_blank(),
      plot.background = ggplot2::element_rect(fill = "#f5f5f2", color = NA),
      panel.background = ggplot2::element_rect(fill = "#f5f5f2", color = NA),
      panel.border = ggplot2::element_blank(),
      plot.title = ggplot2::element_text(size = 16, hjust = 0.5, color = "#4e4d47"),
      plot.subtitle = ggplot2::element_text(
        size = 14,
        hjust = 0.5,
        color = "#4e4d47",
        margin = ggplot2::margin(
          b = -0.1,
          t = -0.1,
          l = 2,
          unit = "cm"
        ),
        debug = F
      ),
      plot.margin = ggplot2::unit(c(.5, .5, .2, .5), "cm"),
      panel.spacing = ggplot2::unit(c(-.1, 0.2, .2, 0.2), "cm"),
      plot.caption = ggplot2::element_text(
        size = 9,
        hjust = 0.92,
        margin = ggplot2::margin(t = 0.2,
                                 b = 0,
                                 unit = "cm"),
        color = "#939184"
      )
    )
}


R/ggtheme.R
#' Bernie's theme for ggplot 2 blog graphs. Inspired by the excellent
#' `{hrbrthemes}` package.
#'
#' @return Theme for ggplot graphs in accordance with Bernie's blog colors.
#' @export
#'
#' @examples
#' \donttest{
#'  library(ggplot2)
#'  # Prepare data
#'  cty_mpg <- aggregate(mpg$cty, by=list(mpg$manufacturer), FUN=mean)  # aggregate
#'  colnames(cty_mpg) <- c("make", "mileage")  # change column names
#'  cty_mpg <- cty_mpg[order(cty_mpg$mileage), ]  # sort
#'  cty_mpg$make <- factor(cty_mpg$make, levels = cty_mpg$make)
#'  # Plot
#'  ggplot(cty_mpg, aes(x=make, y=mileage)) +
#'    geom_segment(aes(x=make,
#'                     xend=make,
#'                     y=0,
#'                     yend=mileage), color = "white") +
#'    geom_point(size=3, color = "orange") +
#'    labs(title="Lollipop Chart",
#'         subtitle="Make Vs Avg. Mileage",
#'         caption="source: mpg") +
#'    HohgantR::themebernie() +
#'    theme(axis.text.x = element_text(angle=65, vjust=0.6))
#'  }

themebernie <- function(){
  hrbrthemes::theme_modern_rc(base_family = "sans",
                              subtitle_family = "sans",
                              caption_family = "sans") +
    ggplot2::theme(panel.background = ggplot2::element_rect(fill = "#222831",
                                                            color = NA),
                   plot.background = ggplot2::element_rect(fill = "#222831",
                                                           color = NA))
}


#' Mineral theme for ggplot 2 graphs
#'
#' @return Theme for ggplot graphs inspired by coal paper and the excellent
#' `{hrbrthemes}` package.
#' @export
#'
#' @examples
#' \donttest{
#'  library(ggplot2)
#'  # Prepare data
#'  cty_mpg <- aggregate(mpg$cty, by=list(mpg$manufacturer), FUN=mean)  # aggregate
#'  colnames(cty_mpg) <- c("make", "mileage")  # change column names
#'  cty_mpg <- cty_mpg[order(cty_mpg$mileage), ]  # sort
#'  cty_mpg$make <- factor(cty_mpg$make, levels = cty_mpg$make)
#'  # Plot
#'  ggplot(cty_mpg, aes(x=make, y=mileage)) +
#'    geom_segment(aes(x=make,
#'                     xend=make,
#'                     y=0,
#'                     yend=mileage), color = "#C3CED6") +
#'    geom_point(size=3, color = "#C3CED6") +
#'    labs(title="Lollipop Chart",
#'         subtitle="Make Vs Avg. Mileage",
#'         caption="source: mpg") +
#'    HohgantR::thememineral() +
#'    theme(axis.text.x = element_text(angle=65, vjust=0.6))
#'  }

thememineral <- function(){
  hrbrthemes::theme_modern_rc(base_family = "sans",
                              subtitle_family = "sans",
                              caption_family = "sans") +
    ggplot2::theme(panel.background = ggplot2::element_rect(fill = "#36454f",
                                                            color = NA),
                   plot.background = ggplot2::element_rect(fill = "#36454f",
                                                           color = NA),
                   panel.grid = ggplot2::element_line(colour = "#718C9E"))
}
